Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,780 @@
+
+		const float k_DirectSurfaceSearchScale = 1.1f;
+		const float k_DirectSurfaceSnapBreakDist = 0.03f;
+		const float k_DirectGroundSnapMin = 0.03f;
+		const float k_DirectGroundSnapMax = 0.07f;
+
+		const float k_WidgetScale = 0.03f;
+
+		const string k_SnappingEnabled = "EditorVR.Snapping.Enabled";
+		const string k_GroundSnapping = "EditorVR.Snapping.Ground";
+		const string k_SurfaceSnapping = "EditorVR.Snapping.Sufrace";
+		const string k_PivotSnapping = "EditorVR.Snapping.Pivot";
+		const string k_SnapRotation = "EditorVR.Snapping.Rotation";
+		const string k_LocalOnly = "EditorVR.Snapping.LocalOnly";
+		const string k_ManipulatorSnapping = "EditorVR.Snapping.Manipulator";
+		const string k_DirectSnapping = "EditorVR.Snapping.Direct";
+
+		const string k_MaterialColorLeftProperty = "_ColorLeft";
+		const string k_MaterialColorRightProperty = "_ColorRight";
+
+		[SerializeField]
+		GameObject m_GroundPlane;
+
+		[SerializeField]
+		GameObject m_Widget;
+
+		[SerializeField]
+		GameObject m_SettingsMenuPrefab;
+
+		[SerializeField]
+		Material m_ButtonHighlightMaterial;
+
+		SnappingModuleSettingsUI m_SnappingModuleSettingsUI;
+		Material m_ButtonHighlightMaterialClone;
+
+		class SnappingState
+		{
+			public Vector3 currentPosition { get; set; }
+			public bool groundSnapping { get; set; }
+			public bool surfaceSnapping { get; set; }
+
+			public Quaternion startRotation { get; private set; }
+			public Bounds identityBounds { get; private set; }
+			public Bounds rotatedBounds { get; private set; }
+
+			public SnappingState(GameObject go, Vector3 position, Quaternion rotation)
+			{
+				currentPosition = position;
+				startRotation = rotation;
+
+				var objTransform = go.transform;
+				var objRotation = objTransform.rotation;
+
+				var rotatedBounds = ObjectUtils.GetBounds(go);
+				go.transform.rotation = Quaternion.identity;
+				var identityBounds = ObjectUtils.GetBounds(go);
+				go.transform.rotation = objRotation;
+
+				rotatedBounds.center -= position;
+				this.rotatedBounds = rotatedBounds;
+				identityBounds.center -= position;
+				this.identityBounds = identityBounds;
+			}
+
+			public SnappingState(GameObject[] objects, Vector3 position, Quaternion rotation)
+			{
+				currentPosition = position;
+				startRotation = rotation;
+
+				var rotatedBounds = ObjectUtils.GetBounds(objects);
+
+				float angle;
+				Vector3 axis;
+				rotation.ToAngleAxis(out angle, out axis);
+				foreach (var obj in objects)
+				{
+					obj.transform.RotateAround(position, axis, -angle);
+				}
+
+				var identityBounds = ObjectUtils.GetBounds(objects);
+
+				foreach (var obj in objects)
+				{
+					obj.transform.RotateAround(position, axis, angle);
+				}
+
+				rotatedBounds.center -= position;
+				this.rotatedBounds = rotatedBounds;
+				identityBounds.center -= position;
+				this.identityBounds = identityBounds;
+			}
+		}
+
+		struct SnappingDirection
+		{
+			public Vector3 direction { get; set; }
```

Comment:
Make these all public fields; According to our convention data in structs can only be fields.